### PR TITLE
allow for configurable server side timeouts on routes

### PR DESCRIPTION
**Pull request by JacobTanenbaum**
using annotations configure a route to have it's own server side timeout

--beginning of route.yaml
apiVersion: v1
kind: Route
metadata:
  annotations:
    hello-again.timeout: 5s
  creationTimestamp: 2016-06-20T22:43:44Z
  name: hello-again
  namespace: default
  resourceVersion: "461544"
  selfLink: /oapi/v1/namespaces/default/routes/hello-again
  uid: 717b8a38-3738-11e6-8b7f-525400d6577c
...

matching against [routename].timeout: [time] is what creates the timeout

### CODE REVIEW

**Comment by JacobTanenbaum** [comment]
@ramr @rajatchopra @knobunc please review?

**Comment by rajatchopra** [comment]
Since the timeout is per route, maybe we should follow the existing convention of the annotation name: router.openshift.io/haproxy.timeout

**Comment by ramr** [comment]
Yeah, agree with @rajatchopra that it would better to just use a standard `"name"` ala: `router.openshift.io/haproxy.timeout` rather than a custom one per route (makes copy-pasta editing easier as well as if you want to change the host name).

And we should also do this for the other `be_tcp_*` and `be_secure_*`  backends.

**Comment by rajatchopra** [comment]
LGTM
@JacobTanenbaum squash the commits maybe.

**Comment by knobunc** [comment]
Is the annotation value validated somewhere?  Or can they sneak in a string and do naughty things?

**Comment by rajatchopra** [comment]
Good point @knobunc 
There is a function call implemented already to check if the value is an integer, we should use that call.

@JacobTanenbaum see https://github.com/openshift/origin/blob/master/pkg/router/template/router.go#L176

**Comment by JacobTanenbaum** [comment]
@rajatchopra @knobunc If I use the isInteger() function then the timeout can only be specified in milliseconds. Most of the other timeouts in the haproxy router template are in seconds, should I make another function for isTime that includes the units so users can input a timeout in any unit that is valid in haproxy?

**Comment by knobunc** [comment]
@rajatchopra added a regex validator... we can just use that.

**Comment by rajatchopra** [comment]
Yes. [0-9]+[su] or something like that.

File: images/router/haproxy/conf/haproxy-config.template [hunk 1]
```diff
@@ -226,6 +226,11 @@ backend be_edge_http_{{$cfgIdx}}
     {{ else }}
   balance leastconn
     {{ end }}
+    {{ with $value := index $cfg.Annotations "router.openshift.io/haproxy.timeout"}}
+      {{if (matchString "[0-9]*(us|ms|s|m|h|d)?" $value) }}
```
Comment (knobunc) [inline images/router/haproxy/conf/haproxy-config.template]: I assume matchString has an implicit anchor to the start and end of the string?  But can we check that "test 9ms test" doesn't match.

Also * matches 0+ characters... I assume you want [0-9]+ to make it match 1+.  Also \d is usually a synonym for [0-9], so let's see if "\d+(us|ms|s|m|h|d)?" works.

Comment (rajatchopra) [inline images/router/haproxy/conf/haproxy-config.template]: 'rm -f *; 90ms test' will match. the regex needs to be ^[1-9][0-9]*(us|ms|s|m|h|d)?$

Comment (knobunc) [inline images/router/haproxy/conf/haproxy-config.template]: Should we make anchoring the front and back the default in matchString?  i.e. wrap the match they give us with ^(...)$

If we want to match a substr, then we can say .*thing.*

It feels like we will bite people with this behavior.

**Comment by rajatchopra** [comment]
Looks good. Squash?

File: pkg/router/template/router.go [hunk 2]
```diff
@@ -187,14 +187,14 @@ func matchValues(s string, allowedValues ...string) bool {
 	return false
 }
 
-func matchString(pattern, s string) bool {
-	glog.V(4).Infof("Matchstring called with %s and %s", pattern, s)
-	status, err := regexp.MatchString(pattern, s)
+func matchPattern(pattern, s string) bool {
+	glog.V(4).Infof("matchPattern called with %s and %s", pattern, s)
+	status, err := regexp.MatchString("^"+pattern+"$", s)
```
Comment (knobunc) [inline pkg/router/template/router.go]: This needs parens, or 'a|b' would match 'anaughty'

**Comment by JacobTanenbaum** [comment]
@knobunc squashed the commits

File: pkg/router/template/plugin.go [hunk 3]
```diff
@@ -99,7 +99,7 @@ func NewTemplatePlugin(cfg TemplatePluginConfig) (*TemplatePlugin, error) {
 	globalFuncs := template.FuncMap{
 		"endpointsForAlias": endpointsForAlias,
 		"env":               env,
-		"matchString":       matchString,
+		"matchPattern":      matchPattern,
```
Comment (knobunc) [inline pkg/router/template/plugin.go]: Are there any docs for this that should change?  I couldn't find any.  So at the least can we put a comment over in the original PR that added matchString to say that it changed (and what the new behavior is).

File: pkg/router/template/router.go [hunk 4]
```diff
@@ -187,14 +187,14 @@ func matchValues(s string, allowedValues ...string) bool {
...
 }
 
-func matchString(pattern, s string) bool {
-	glog.V(4).Infof("Matchstring called with %s and %s", pattern, s)
-	status, err := regexp.MatchString(pattern, s)
+func matchPattern(pattern, s string) bool {
+	glog.V(4).Infof("matchPattern called with %s and %s", pattern, s)
+	status, err := regexp.MatchString("^("+pattern+")$", s)
 	if err == nil {
 		glog.V(4).Infof("Matchstring returning status: %v", status)
```
Comment (knobunc) [inline pkg/router/template/router.go]: This should change to matchPattern

File: pkg/router/template/plugin.go [hunk 5]
```diff
@@ -99,7 +99,7 @@ func NewTemplatePlugin(cfg TemplatePluginConfig) (*TemplatePlugin, error) {
 	globalFuncs := template.FuncMap{
```
Comment (knobunc) [inline pkg/router/template/plugin.go]: We need to doc what these do... can we at least put a one-line comment on them to summarize usage?

**Comment by knobunc** [comment]
LGTM [merge]

**Pull request merged**